quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -444,67 +545,45 @@ impl IngestController {
+            .iter()
+            .map(|ingester| (ingester.as_str(), 0))
+            .collect();
 
         for shard in model.all_shards() {
             if shard.is_open() && !unavailable_leaders.contains(&shard.leader_id) {
-                num_open_shards += 1;
-
-                *per_leader_num_open_shards
-                    .entry(&shard.leader_id)
-                    .or_default() += 1;
-            }
-        }
-        let mut num_remaining_shards_to_allocate = num_shards_to_allocate;
-        let num_open_shards_target = num_shards_to_allocate + num_open_shards;
-        let max_num_shards_to_allocate_per_node = num_open_shards_target / num_ingesters;
-
-        // Allocate at most `max_num_shards_to_allocate_per_node` shards to each ingester.
-        for (leader_id, follower_id) in ingesters.iter().zip(ingesters.iter().cycle().skip(1)) {
-            if num_remaining_shards_to_allocate == 0 {
-                break;
-            }
-            let num_open_shards_inner = per_leader_num_open_shards
-                .get(leader_id.as_str())
-                .copied()
-                .unwrap_or_default();
-
-            let num_shards_to_allocate_inner = max_num_shards_to_allocate_per_node
-                .saturating_sub(num_open_shards_inner)
-                .min(num_remaining_shards_to_allocate);
-
-            for _ in 0..num_shards_to_allocate_inner {
-                num_remaining_shards_to_allocate -= 1;
-
-                let leader = leader_id.clone();
-                let mut follower_opt = None;
-
-                if self.replication_factor > 1 {
-                    follower_opt = Some(follower_id.clone());
+                if let Some(shard_count) =
+                    per_node_num_open_shards.get_mut(shard.leader_id.as_str())
+                {
+                    *shard_count += 1;
+                }
+                if let Some(follower_id) = shard.follower_id.as_ref() {
+                    if let Some(shard_count) =
+                        per_node_num_open_shards.get_mut(follower_id.as_str())
+                    {
+                        *shard_count += 1;
+                    }
                 }
-                leader_follower_pairs.push((leader, follower_opt));
             }
         }
-        // Allocate remaining shards one by one.
-        for (leader_id, follower_id) in ingesters.iter().zip(ingesters.iter().cycle().skip(1)) {
-            if num_remaining_shards_to_allocate == 0 {
-                break;
-            }
-            num_remaining_shards_to_allocate -= 1;
 
-            let leader = leader_id.clone();
-            let mut follower_opt = None;
-
-            if self.replication_factor > 1 {
-                follower_opt = Some(follower_id.clone());
-            }
-            leader_follower_pairs.push((leader, follower_opt));
-        }
-        Some(leader_follower_pairs)
+        assert!(self.replication_factor == 1 || self.replication_factor == 2);
```

Comment:
why do we do this assert now? can we ensure this before? like when we build the controller?
Adding a comment on the variable to mention that would be nice too.

Reply:
It is enforced before. The assert is here to be defensive and as documentation. An enum or a bool would have been nice.